sdk/core/azure_core/src/http/policies/bearer_token_policy.rs

```diff
@@ -59,24 +70,143 @@ impl Policy for BearerTokenAuthorizationPolicy {
+                            stream.reset().await?;
+                        }
+                        response = next[0].send(ctx, request, &next[1..]).await?
+                    }
+                }
+            }
+        }
+
+        Ok(response)
+    }
+}
+
+/// Callback [`BearerTokenAuthorizationPolicy`] invokes when it receives a 401 Unauthorized response with an authentication challenge (WWW-Authenticate header).
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait OnChallenge: std::fmt::Debug + Send + Sync {
+    /// Called when [`BearerTokenAuthorizationPolicy`] receives a 401 Unauthorized response with a challenge. Implementations are
+    /// responsible for parsing authentication parameters from the challenge, authorizing the request via the provided [`Authorizer`],
+    /// and indicating whether the policy should retry the request.
+    ///
+    /// # Arguments
+    /// * `context` - The request context
+    /// * `request` - The HTTP request that received the challenge
+    /// * `authorizer` - Helper used to acquire an access token and set the request's authorization header
+    /// * `headers` - The 401 response's headers
+    ///
+    /// # Returns
+    /// * `Ok(true)` when the callback handled the challenge and [`BearerTokenAuthorizationPolicy`] should retry the request.
+    /// * `Ok(false)` when the callback can't handle the challenge. [`BearerTokenAuthorizationPolicy`] will return the 401 response to the client in this case.
+    /// * `Err` when an error occurs while handling the challenge.
+    async fn on_challenge(
+        &self,
+        context: &Context,
+        request: &mut Request,
+        authorizer: &dyn Authorizer,
+        headers: &Headers,
+    ) -> Result<bool>;
+}
+
+/// Callback [`BearerTokenAuthorizationPolicy`] invokes on every request it receives, before sending the request.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait OnRequest: std::fmt::Debug + Send + Sync {
+    /// Invoked once on every [`BearerTokenAuthorizationPolicy::send`] invocation, before the policy sends the request.
+    /// `send` doesn't call this method before retrying a request after an authentication challenge (see [`OnChallenge`]
+    /// for more about challenge handling). Implementations are responsible for authorizing each request via the provided
+    /// [`Authorizer`]. The policy sends the request when this method returns Ok.
+    ///
+    /// # Arguments
+    /// * `context` - The request context
+    /// * `request` - The HTTP request being sent
+    /// * `authorizer` - Helper used to acquire an access token and set the request's authorization header
+    async fn on_request(
+        &self,
+        context: &Context,
+        request: &mut Request,
+        authorizer: &dyn Authorizer,
+    ) -> Result<()>;
+}
+
+mod sealed {
+    pub trait Sealed {}
+}
```

Comment:
Don't define your own. Just use `crate::private::Sealed`. Ancestors' private members are visible to all descendants.